### ISSUE TITLE
Add Autocomplete feature

### ISSUE THREAD

**Issue by aksonov** [decosoftware/deco-ide#40]
For regular JS classes as well for React Components.

**Comment by dabbott** [comment]
At a high level, what's involved here:

In a separate thread, index all the files in a project with a static analysis engine.

Options for engines:
- flow
- typescript
- ternjs
- others?

Probably flow makes most sense, since its specifically for RN. However, typescript does seem to work spectacularly in VSCode, so worth looking into.

As you type, re-index the current file and look up type info.

A large task, but probably a lot of stuff could be taken from Nuclide or VSCode. If somebody wants to help me tackle this it would be cool :)

**Comment by DanielRosenwasser** [comment]
Hey @dabbott, one of the nice things is that with TS integration users would have the choice of both JS and TS, as well as an easier cross-platform story for the near future. I'm actually on the TypeScript team if you're interested, so if you had any questions regarding this, I'd love to help out. :smiley:

**Comment by dabbott** [comment]
@DanielRosenwasser awesome, thanks :)

How does TS work with Flow code? I see quite a lot of Flow type annotations in RN apps, but not much TS. If TS doesn't handle Flow, I'd probably choose Flow based on its popularity (and being what FB uses internally and in RN core). I don't know either very well, but the type annotation syntax looks pretty similar.

Does TS use babel or support babel extensions? If not, I'd worry about language support lagging behind for things like ES7 decorators, static class properties, async functions, etc.

**Comment by DanielRosenwasser** [comment]
> ES7 decorators, static class properties, async functions

Those first two things were actually originally first designed/implemented in TypeScript, and we currently support parsing all of them. :smile: 

> Does TS use babel or support babel extensions?

TypeScript itself doesn't use Babel, but it fits well into a pipeline with it.

> If not, I'd worry about language support lagging

Keep in mind that TS itself does't necessarily need to come into the pipeline at all. For completion, you're only concerned with the editing scenario, right? For instance, in VS Code, the JS language service is powered by TypeScript, but there's no compile step there. You can let Babel do the heavy lifting.

Your concern about future features is a fair point, but Flow actually has the same problem - it has its own parser that is independent of Babel. We also generally don't implement proposals unless they are fairly stable and mature. From the time a feature is proposed, it can change drastically, or even get outright rejected at later stages. For instance, callable constructors were supported in Babel by the `classConstructorCall` option, but [it is no longer an active proposal](https://github.com/tc39/proposals/blob/83d709af67eb1ddcbe962283ed0864ae051bc0d2/inactive-proposals.md). We try to avoid leaving users in those situations when possible.

**Comment by dabbott** [comment]
@DanielRosenwasser Very interesting! I didn't know that Flow had a separate, non-babel parser. I saw babel in the package.json and jumped to conclusions.

Does TS handle Flow type annotations (are they the same as TS type annotations?)?

**Comment by dabbott** [comment]
Added Flow, mainly for the AST. Flow always builds a full parse tree, even if there are errors in the code, which is what we need for Deco. Flow also does autocomplete - not very well though. Maybe revisit using TS language services at some point.

**Comment by DanielRosenwasser** [comment]
Hey @dabbott, sorry I missed the preceding message. There are a lot of similarities between TypeScript syntax and Flow syntax. For most places there's a lot of overlap. I would say in terms of core syntax, TS supports a superset, but differs in a few places like using nullables (`?T` vs `T | null`).

Also, totally respect your decision. Having worked on our compiler and language service though, I know that we have specifically worked to ensure that we recover gracefully and give a good experience even in the face of errors. Since you brought that up, was that not the experience you saw when you investigated TypeScript?

**Comment by dabbott** [comment]
Hey @DanielRosenwasser. I needed a full AST for our upcoming release, so that was my main goal. I first tried Babel, which would be my #1 choice for parser, given that it's used by the RN packager. However, Babel couldn't recover gracefully. Flow seemed super easy to integrate (`npm install flow-bin`, and then send it commands) and I've been very happy with the AST quality/speed. I didn't get around to trying TypeScript.

The main difficulty when I investigated TypeScript (back in June) was the documentation. It wasn't clear how to integrate TS, and the VSCode source is a bit difficult to follow. Do you have an example project, or can you give a few steps I can try to get autocompletion on vanilla JS?

I don't think it has to be one or the other necessarily. The VSCode autocomplete is definitely the best I've seen for RN, so it might make sense to use both. I'm very disappointed with Flow in terms of the actual type information I can get out of it. Hardly any types, and a very limited API.

**Comment by DanielRosenwasser** [comment]
Yeah, I can see why that'd be a problem. Apart from VS Code, here's some other editor support that uses TS:

* Our [Sublime Text Plugin](https://github.com/Microsoft/TypeScript-Sublime-Plugin) (in Python)
* [Tsuquyomi](https://github.com/Quramy/tsuquyomi) (a Vim plugin)
* [tide](https://github.com/ananthakumaran/tide) (an Emacs plugin)
* [typescript.java](https://github.com/angelozerr/typescript.java) (an Eclipse plugin).

The commands that are supported are available [here](https://github.com/Microsoft/TypeScript/blob/master/src/server/session.ts#L102-L130). The actual data that you can send to/get from the server is defined [here](https://github.com/Microsoft/TypeScript/blob/3cca17e17d77224d3de2064fad26e0a339e237e2/src/server/session.ts).

So if we took a look at [`CompletionRequest`s](https://github.com/Microsoft/TypeScript/blob/3cca17e17d77224d3de2064fad26e0a339e237e2/src/server/protocol.d.ts#L710-L718), we can see basically a `Request` where the `command` field is `"completions"`. Its `arguments` field is a `CompletionsRequestArgs`, which has an optional prefix to narrow completions on, along with whatever fields a `FileLocationRequestArgs` has. So the request would look like this:

```json
{
    "seq": 100,
    "type": "request",
    "command": "completions",
    "arguments": {
        "line": 1,
        "offset": 14,
        "prefix": "puppie",
    }
}
```

I'll be away from the office for most of the month, but I think I'll try to figure something out with @mhegazy and @billti about documenting TSServer a little better in the near future.
